pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -463,4 +541,57 @@ private long getWatermarkMs(long minStartTimeMs, long bucketMs, String mergeLeve
 
     return pinotTaskConfigs;
   }
+
+  private long getMergeRollupTaskDelayInNumTimeBuckets(long watermarkMs, long bufferTimeMs, long bucketTimeMs) {
+    if (bufferTimeMs == 0) {
+      return 0;
+    }
+    return (long) Math.floor((System.currentTimeMillis() - watermarkMs - bufferTimeMs) / (double) bucketTimeMs);
+  }
+
+  private void setWatermarkMs(String offlineTableName, String mergeLevel, long watermarkMs, long bufferTimeMs,
+      long bucketTimeMs) {
+    LOGGER.info(
+        "Setting wartermark for table: {} and mergeLevel: {} is {} (watermarkMs={}, bufferTimeMs={}, bucketTimeMs={})",
+        offlineTableName, mergeLevel, getMergeRollupTaskDelayInNumTimeBuckets(watermarkMs, bufferTimeMs, bucketTimeMs),
+        watermarkMs, bucketTimeMs, bucketTimeMs);
+
+    ControllerMetrics controllerMetrics = _clusterInfoAccessor.getControllerMetrics();
+    PinotMetricsRegistry metricsRegistry = controllerMetrics.getMetricsRegistry();
+    if (metricsRegistry == null) {
+      return;
+    }
+
+    // Update gauge value that indicates the delay in terms of the number of time buckets.
+    MetricKey metricKey = new MetricKey(offlineTableName, mergeLevel);
+    _mergeRollupWatermarks.compute(metricKey, (k, v) -> {
+      if (v == null) {
+        controllerMetrics.addCallbackGaugeIfNeeded(getMetricNameForTaskDelay(metricKey.toString()),
+            (() -> getMergeRollupTaskDelayInNumTimeBuckets(_mergeRollupWatermarks.getOrDefault(k, 0L), bufferTimeMs,
+                bucketTimeMs)));
+      }
+      return watermarkMs;
+    });
+  }
+
+  private void resetWatermarkMs(String tableNameWithType) {
+    ControllerMetrics controllerMetrics = _clusterInfoAccessor.getControllerMetrics();
+    PinotMetricsRegistry metricsRegistry = controllerMetrics.getMetricsRegistry();
+    if (metricsRegistry == null) {
+      return;
+    }
+    Iterator<MetricKey> it = _mergeRollupWatermarks.keySet().iterator();
+    while (it.hasNext()) {
+      MetricKey metricKey = it.next();
+      if (metricKey.getTableNameWithType().equals(tableNameWithType)) {
+        it.remove();
+        controllerMetrics.removeCallbackGauge(getMetricNameForTaskDelay(metricKey.toString()));
```

Comment:
Can we simply return if this if statement is qualified? By doing so, you don't have to traverse the whole map.

Reply:
There can be multiple `MetricKey` with the same `tableName`. We will have 1 entry per `<tableName, mergeLevel>` pair. So, we will need to traverse the entire map...

Reply:
How about introducing an extra map to store the table name and its list of `MetricKey`s? By doing so, you don't have to traverse the whole map. And you can also introduce a new class so that the logic of adding/removing metrics can be encapsulated.

Reply:
@jackjlli I changed the watermark map as 2 level map (tableName -> mergeLevel -> watermark). Can you go over one more time?